File: android/src/main/kotlin/com/onlinepayments/client/flutter/online_payments_sdk/sdk/models/ResultError.kt
/*
 * Do not remove or alter the notices in this preamble.
 *
 * This software is owned by Worldline and may not be be altered, copied, reproduced, republished, uploaded, posted, transmitted or distributed in any way, without the prior written consent of Worldline.
 *
 * Copyright © 2025 Worldline and/or its affiliates.
 *
 * All rights reserved. License grant and user rights and obligations according to the applicable license agreement.
 *
 * Please contact Worldline for questions regarding license and user rights.
 */

package com.onlinepayments.client.flutter.online_payments_sdk.sdk.models

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.security.InvalidParameterException

data class ResultError(val code : String, val message: String) {
    companion object {
        enum class ResultErrorCode(val id: String) {
            NOT_INITIALIZED("1"),
            MISSING_REQUEST_ARGUMENT("2"),
            MISSING_REQUIRED_ARGUMENT("3"),
            MISSING_CONTEXT("4"),
        }
        fun missingRequestArgument(call: MethodCall, result: MethodChannel.Result, e: Exception?) {
            return result.error(ResultErrorCode.MISSING_REQUEST_ARGUMENT.id, "Missing the required request argument for request : ${call.method}", e)
        }

        fun notInitialized(result: MethodChannel.Result) {
            return result.error(ResultErrorCode.NOT_INITIALIZED.id, "Session is null, did you forget to initialize it?", null)
        }

        fun missingRequiredArgument(call: MethodCall, result: MethodChannel.Result, e: InvalidParameterException? = null) {
            return result.error(ResultErrorCode.MISSING_REQUIRED_ARGUMENT.id, "Missing a required argument in the request for method : ${call.method}", e)
        }

        fun missingContext(call: MethodCall, result: MethodChannel.Result) {
            return result.error(ResultErrorCode.MISSING_CONTEXT.id, "No context available to execute method : ${call.method}", null)
        }
    }
}

File: ios/Classes/Errors/ResultError.swift
/*
 * Do not remove or alter the notices in this preamble.
 *
 * This software is owned by Worldline and may not be be altered, copied, reproduced, republished, uploaded, posted, transmitted or distributed in any way, without the prior written consent of Worldline.
 *
 * Copyright © 2025 Worldline and/or its affiliates.
 *
 * All rights reserved. License grant and user rights and obligations according to the applicable license agreement.
 *
 * Please contact Worldline for questions regarding license and user rights.
 */
import Flutter

struct ResultError: Encodable {
    let code: String
    let message: String

    static func sessionNotInitialized() -> FlutterError {
        return FlutterError(code: ResultErrorCode.sessionNotInitialized.rawValue, message: "Session is nil. Please initialize the session first", details: nil)
    }

    static func invalidParameter(parameter: String) -> FlutterError {
        return FlutterError(code: ResultErrorCode.invalidParameter.rawValue, message: "The following parameter is invalid : \(parameter)", details: nil)
    }

    static func dataError<T: Decodable>(object: T.Type, call: FlutterMethodCall) -> FlutterError {
        return FlutterError(code: ResultErrorCode.dataError.rawValue, message: "Could not create Data object of \(object.self) for request : \(call.method)", details: nil)
    }

    static func decodingError<T: Decodable>(object: T.Type, call: FlutterMethodCall) -> FlutterError {
        return FlutterError(code: ResultErrorCode.decodingError.rawValue, message: "Could not decode \(object.self) for request : \(call.method)", details: nil)
    }

    static func encodingError(type: EncodingErrorType) -> FlutterError {
        return FlutterError(code: ResultErrorCode.encodingError.rawValue, message: "Could not encode \(type) to String", details: nil)
    }
}

enum ResultErrorCode: String {
    case sessionNotInitialized = "1"
    case invalidParameter = "2"
    case dataError = "3"
    case decodingError = "4"
    case encodingError = "5"
}

enum EncodingErrorType {
    case success
    case failure
    case apiFailure
    case maskedValuesDictionary
    case unmaskedValuesDictionary
    case validationErrors
}
